### PR TITLE
Extend order with subtotal

### PR DESCRIPTION
**Pull request by michal-macioszczyk**
I want to merge this change because I want to be able to have subtotal field on `Order` instance. 

[14653](https://github.com/saleor/saleor/issues/14653
)

PR for [3.18](https://github.com/saleor/saleor/pull/14932) version 

# Impact

- [x] New migrations
- [ ] New/Updated API fields or mutations
- [ ] Deprecated API fields or mutations
- [ ] Removed API types, fields, or mutations

# Docs

- [ ] Link to documentation:

# Pull Request Checklist

- [ ] Privileged queries and mutations are either absent or guarded by proper permission checks
- [ ] Database queries are optimized and the number of queries is constant
- [ ] Database migrations are either absent or optimized for zero downtime
- [ ] The changes are covered by test cases
- [ ] All new fields/inputs/mutations have proper labels added (`ADDED_IN_X`, `PREVIEW_FEATURE`, etc.)
- [ ] All migrations have proper dependencies
- [ ] All indexes are added concurrently in migrations
- [ ] All RunSql and RunPython migrations have revert option defined

### CODE REVIEW

File: saleor/order/calculations.py [hunk 1]
```diff
@@ -499,8 +502,7 @@ def order_subtotal(
     order, lines = fetch_order_prices_if_expired(order, manager, lines, force_update)
     # Lines aren't returned only if
     # we don't pass them to `fetch_order_prices_if_expired`.
-    lines = cast(Iterable[OrderLine], lines)
-    return quantize_price(get_subtotal(lines, currency), currency)
+    return quantize_price(order.subtotal, currency)
```
[the remark this answers is not in the document]
Comment (korycins) [reply saleor/order/calculations.py:505]: For sure, we can't use this field until it is fully propagated. 
So as you said:
3.18.X receives the propagation logic.
3.19 receives the logic which will use these data.
You could also do:
3.19 receives the propagation logic
3.20 receives the logic which will use these data

File: saleor/order/migrations/tasks/saleor3_19.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+from ...models import Order, OrderLine
+
+# celery task for one batch took 0.1s and 10.54MB at its peak
+BATCH_SIZE = 1000
+
+
+@app.task
+def update_order_subtotals(batch_id=None, batch_size=BATCH_SIZE):
+    # Determine the starting and ending order numbers for the current batch
+    if batch_id is None:
```
[the remark this answers is not in the document]
Comment (michal-macioszczyk) [reply saleor/order/migrations/tasks/saleor3_19.py:13]: done

File: saleor/checkout/complete_checkout.py [hunk 3]
```diff
@@ -1234,8 +1240,17 @@ def _create_order_from_checkout(
     # giftcards
     currency = checkout_info.checkout.currency
     subtotal_list = [line.line.total_price for line in order_lines_info]
-    subtotal = sum(subtotal_list, zero_taxed_money(currency))
-    total_without_giftcard = subtotal + shipping_total - checkout_info.checkout.discount
+    order.subtotal = sum(subtotal_list, zero_taxed_money(currency))
```
[the remark this answers is not in the document]
Comment (michal-macioszczyk) [reply saleor/checkout/complete_checkout.py:1243]: fixed

File: saleor/order/models.py [hunk 4]
```diff
@@ -412,7 +425,7 @@ def is_shipping_required(self):
         return any(line.is_shipping_required for line in self.lines.all())
 
     def get_subtotal(self):
-        return get_subtotal(self.lines.all(), self.currency)
+        return self.subtotal
```
[the remark this answers is not in the document]
Comment (michal-macioszczyk) [reply saleor/order/models.py]: I deleted it

File: saleor/order/calculations.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (michal-macioszczyk) [reply saleor/order/calculations.py:505]: done

File: saleor/order/migrations/0181_order_subtotal_as_a_field.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+from django.apps import apps as registry
+from django.db import migrations
+from django.db.models.signals import post_migrate
+
+from .tasks.saleor3_19 import update_order_subtotals
+
+
+def trigger_update_order_subtotals(apps, _schema_editor):
+    def on_migrations_complete(sender=None, **kwargs):
+        update_order_subtotals.delay()
```
Comment (korycins) [inline saleor/order/migrations/0181_order_subtotal_as_a_field.py:10]: For main, the celery task should be converted to normal migration. The celery task is only valid for the migration that is placed in released version

File: saleor/order/migrations/tasks/saleor3_19.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
+from django.db.models import F, Func, Min, OuterRef, Subquery
```
Comment (korycins) [inline saleor/order/migrations/tasks/saleor3_19.py:1]: This should be converted to normal migration

File: saleor/order/migrations/0181_order_subtotal_as_a_field.py (hunk 5 above)
Comment (michal-macioszczyk) [reply saleor/order/migrations/0181_order_subtotal_as_a_field.py]: done

**Review by jakubkuc — APPROVED** [review]

File: saleor/order/migrations/0181_order_subtotal_as_a_field.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    if first_order is None:
+        return  # Exit if there are no orders
+    start_number = first_order.number
+
+    while True:
+        # right here we need to find only Orders which was not populated
+        # with subtotal in previous release, so the amount of rows to populate
+        # shouldn't be big
+        queryset = Order.objects.order_by("number").filter(
+            number__gte=start_number, subtotal_gross_amount__exact=0
```
Comment (IKarbowiak) [inline saleor/order/migrations/0181_order_subtotal_as_a_field.py:22]: I've just realized that when there is an Order without any lines this method will be run indefinitely

**Review by IKarbowiak — APPROVED** [review]

File: saleor/order/migrations/0181_order_subtotal_as_a_field.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        orders_to_update = []
+        for order in orders_with_totals:
+            order.subtotal_net_amount = order.subtotal_net_sum
+            order.subtotal_gross_amount = order.subtotal_gross_sum
+            orders_to_update.append(order)
+
+        Order.objects.bulk_update(
+            orders_to_update, ["subtotal_net_amount", "subtotal_gross_amount"]
+        )
+        start_number = current_batch_order_numbers[-1]
```
Comment (korycins) [inline saleor/order/migrations/0181_order_subtotal_as_a_field.py:55]: #14932, in celery task you have:
```python
last_number = current_batch_order_numbers[-1]
update_order_subtotals.delay(last_number + 1)
```
I don't see `+1` here. Which one is correct?

**Review by korycins — APPROVED** [review]

File: saleor/order/migrations/0181_order_subtotal_as_a_field.py (hunk 8 above)
Comment (michal-macioszczyk) [reply saleor/order/migrations/0181_order_subtotal_as_a_field.py:55]: Both are correct, first one in `3.18` version is Celery based and the `main` one is non-Celery.

Non-Celery: The continuous loop retains context, and the `gte` query implicitly moves to the next unprocessed order.

Celery: Each task is independent, requiring an explicit definition of the starting point for each new task `(last_number + 1)`, ensuring no overlap and that each order is processed only once.

**Review by korycins — APPROVED** [review]

**Pull request merged**